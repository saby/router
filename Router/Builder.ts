/**
 * Рендеринг статичной страницы в билдере
 * @library
 * @private
 * @author Мустафин Л.И.
 * @module
 */

export { TIMETESTER_SCRIPTS_NAMESPACE } from './_Builder/_Bootstrap/DataAggregators/UtilsScripts';
export { IRenderOptions, IPageConfig } from './_Builder/_Bootstrap/Interface';
export { mainRender, renderHTMLforOldRoutes } from './_Builder/Bootstrap';
export { renderStatic, IRenderBuilderOptions } from './_Builder/Static';
