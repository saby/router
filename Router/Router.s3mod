<?xml version="1.0" encoding="UTF-8" ?>
<ui_module for_cdn="0" id="a857d050-c754-4d39-90ec-fca5e2ddf022" name="Router" kaizen_zone="4ffaea04-d83a-4b80-ad28-18d4ab9c3e06" package="Wasaby" required="1" autodoc="1" responsible="Мустафин Л.И." responsible_uuid="11f1d1c5-8d45-4bc0-bcd7-9745bf8ddebc" version="1.00">

  <description>Библиотека, предназначенная для определения правил построения приложения в зависимости от URL запроса. Решает задачи построения страницы на сервере и организации SPA-переходов при навигации на клиенте.</description>

  <depends>
    <module id="0f5d26db-dd61-4957-9375-8861ad66f219" name="Vdom"/>
    <module id="10efeac7-0f02-4c73-9474-0d1d0db077b6" name="View"/>
    <module id="323fd54f-1fab-42a6-8aea-6726d6392d53" name="Application"/>
    <module id="636177e5-4d75-44fa-9173-977dc90ff111" name="Env"/>
    <module id="686f183d-0bcc-4732-b4e4-6016c2a67a0f" name="Types"/>
    <module id="a5e32491-7b02-45b1-b65d-53e9723ec0cd" name="React"/>
    <module id="f27abe12-19df-4c9f-b886-634bd20a0011" name="UI"/>
    <module id="1456f36f-76ef-4323-a2dd-a22bc91e7209" name="Compiler"/>
  </depends>

  <load_after/>

  <parameters/>

  <resources/>

  <preload/>

  <typescript>
    <arg key="typecheck" value="1"/>
  </typescript>
</ui_module>
