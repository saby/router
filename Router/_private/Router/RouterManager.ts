import {
    TStateChangeFunction,
    IHistoryState,
    IRegisteredRoute,
    IRegisteredReference,
    IRegisterableComponent,
} from '../DataInterfaces';

/**
 * Интерфейс менеджера обработчиков (до и после SPA перехода) для контролов Route и Reference.
 * @public
 * @author Мустафин Л.И.
 */
export interface IRouterManager {
    /**
     * Добавление обработчиков до и после SPA перехода для контрола Route.
     * @hidden
     */
    addRoute(
        route: IRegisterableComponent,
        beforeUrlChangeCb: TStateChangeFunction,
        afterUrlChangeCb: TStateChangeFunction
    ): void;
    /**
     * Удаление обработчиков до и после SPA перехода для контрола Route.
     */
    removeRoute(route: IRegisterableComponent): void;
    /**
     * Добавление обработчика после SPA перехода для контрола Reference.
     */
    addReference(
        reference: IRegisterableComponent,
        afterUrlChangeCb: TStateChangeFunction
    ): void;
    /**
     * Добавление обработчика после SPA перехода для контрола Reference.
     */
    removeReference(reference: IRegisterableComponent): void;
    /**
     * Вызов обработчиков до SPA перехода для контролов Route.
     */
    callBeforeUrlChange(
        newState: IHistoryState,
        currentState: IHistoryState
    ): Promise<boolean> | boolean;
    /**
     * Вызов обработчиков после SPA перехода для контролов Route и Reference.
     */
    callAfterUrlChange(newState: IHistoryState, oldState: IHistoryState): void;
}

/**
 * Менеджер обработчиков (до и после SPA перехода) для контролов Route и Reference.
 * @private
 */
export default class RouterManager implements IRouterManager {
    protected _registeredRoutes: Record<string, IRegisteredRoute> = {};
    protected _registeredReferences: Record<string, IRegisteredReference> = {};

    addRoute(
        route: IRegisterableComponent,
        beforeUrlChangeCb: TStateChangeFunction,
        afterUrlChangeCb: TStateChangeFunction
    ): void {
        this._registeredRoutes[route.getInstanceId()] = {
            beforeUrlChangeCb,
            afterUrlChangeCb,
        };
    }

    removeRoute(route: IRegisterableComponent): void {
        delete this._registeredRoutes[route.getInstanceId()];
    }

    addReference(
        reference: IRegisterableComponent,
        afterUrlChangeCb: TStateChangeFunction
    ): void {
        this._registeredReferences[reference.getInstanceId()] = {
            afterUrlChangeCb,
        };
    }

    removeReference(reference: IRegisterableComponent): void {
        delete this._registeredReferences[reference.getInstanceId()];
    }

    callBeforeUrlChange(
        newState: IHistoryState,
        currentState: IHistoryState
    ): Promise<boolean> | boolean {
        // Если нет обработчиков для route'ов, то не меняем текущий url
        if (!Object.keys(this._registeredRoutes).length) {
            return false;
        }

        // вызовы АСИНХРОННЫХ предобработчиков смены url-адреса
        let callbackPromises: Promise<boolean>[] = [];
        // вызовы СИНХРОННЫХ предобработчиков смены url-адреса
        const callbackResults: boolean[] = [];

        for (const routeId in this._registeredRoutes) {
            if (!this._registeredRoutes.hasOwnProperty(routeId)) {
                continue;
            }
            const route: IRegisteredRoute = this._registeredRoutes[routeId];
            const beforeCb: Promise<boolean> | boolean =
                route.beforeUrlChangeCb(newState, currentState);
            if (beforeCb instanceof Promise) {
                callbackPromises.push(beforeCb);
            } else {
                callbackResults.push(beforeCb);
            }
        }

        // АСИНХРОННЫЙ результат выполнения предобработчиков смены url-адреса
        if (callbackPromises.length) {
            callbackPromises = callbackPromises.concat(
                callbackResults.map((r) => {
                    return Promise.resolve(r);
                })
            );
            return Promise.all(callbackPromises).then((results) => {
                return results.indexOf(false) === -1;
            });
        }

        // СИНХРОННЫЙ результат выполнения предобработчиков смены url-адреса
        return callbackResults.indexOf(false) === -1;
    }

    callAfterUrlChange(newState: IHistoryState, oldState: IHistoryState): void {
        for (const routeId in this._registeredRoutes) {
            if (this._registeredRoutes.hasOwnProperty(routeId)) {
                this._registeredRoutes[routeId].afterUrlChangeCb(
                    newState,
                    oldState
                );
            }
        }

        for (const referenceId in this._registeredReferences) {
            if (this._registeredReferences.hasOwnProperty(referenceId)) {
                this._registeredReferences[referenceId].afterUrlChangeCb(
                    newState,
                    oldState
                );
            }
        }
    }
}
