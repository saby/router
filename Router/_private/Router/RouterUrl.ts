import { getConfig, ILocation } from 'Application/Env';
import UrlRewriter from '../UrlRewriter';

/**
 * Интерфейс класса работы с url в рамках Роутинга.
 * @public
 * @author Мустафин Л.И.
 */
export interface IRouterUrl {
    /**
     * window.location или его реализация в случае открытия приложения на панели.
     */
    readonly location: ILocation;
    /**
     * Возвращает url внутреннего состояния Роутера.
     */
    getStateUrl(): string;
    /**
     * Установка нового актуального значения url'а внутреннего состояния Роутера.
     * @hidden
     */
    setStateUrl(value: string): void;
    /**
     * Возвращает url из адресной строки.
     */
    getUrl(): string;
    /**
     * Получить url с префиксом сервиса вначале.
     * Актуален на СП, т.к. для стороннего сервиса на СП поле appRoot === '/', тогда как на клиенте appRoot = '/external'
     * напр. находимся на странице /page/some стороннего сервиса /external
     * Тогда этот метод на СП вернет строчку /external/page/some
     * Когда метод getStateUrl на СП вернет строчку вида /page/some
     */
    getServiceUrl(url?: string): string;
}
/**
 * Класс для работы с url в рамках Роутинга.
 * @private
 * @author Мустафин Л.И.
 */
export default class RouterUrl implements IRouterUrl {
    /**
     * Url внутреннего состояния Роутера
     */
    private _stateUrl: string;

    constructor(public location: ILocation, urlRewriter: UrlRewriter) {
        this._stateUrl = urlRewriter.get(this.getUrl());
    }

    /**
     * Возвращает url внутреннего состояния Роутера.
     */
    getStateUrl(): string {
        return this._stateUrl || this.getUrl();
    }

    /**
     * Установка нового актуального значения url'а внутреннего состояния Роутера.
     */
    setStateUrl(value: string): void {
        this._stateUrl = value;
    }

    /**
     * Возвращает url из адресной строки.
     */
    getUrl(): string {
        return this.location.pathname + this.location.search + this.location.hash;
    }

    /**
     * Получить url с префиксом сервиса вначале.
     * Актуален на СП, т.к. для стороннего сервиса на СП поле appRoot === '/', тогда как на клиенте appRoot = '/external'
     * напр. находимся на странице /page/some стороннего сервиса /external
     * Тогда этот метод на СП вернет строчку /external/page/some
     * Когда метод getStateUrl на СП вернет строчку вида /page/some
     */
    getServiceUrl(url?: string): string {
        const href = url || this.getUrl();
        let appRoot: string = getConfig('appRoot');

        // если не задан appRoot или это "основной" сервис
        if (!appRoot || appRoot === '/') {
            return href;
        }

        appRoot = appRoot.replace(/\//g, '');
        const hrefParts = href.replace(/^\//, '').split('/');

        // если в начале url НЕТ "названия" сервиса
        // то нужно добавмить его в начало url
        if (hrefParts[0] !== appRoot) {
            hrefParts.unshift(appRoot);
        }

        return '/' + hrefParts.join('/');
    }
}
