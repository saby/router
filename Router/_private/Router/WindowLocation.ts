import { ILocation } from 'Application/Interface';
import { UrlParts } from '../MaskResolver/UrlParts';

/**
 * Интерфейс заглушки-location для создания внутри роутера при работе роутера в приложении на панели
 * @private
 */
export interface IWindowLocation extends ILocation {
    _update(url: string): void;
}

/**
 * Класс-заглушка для window.location внутри роутера при работе роутера в приложении на панели
 * @private
 */
export default class WindowLocation implements IWindowLocation {
    protocol: string = '';
    host: string = '';
    hostname: string = '';
    port: string = '';
    href: string = '';
    pathname: string = '';
    search: string = '';
    hash: string = '';

    constructor(initialUri: string = '/') {
        this._update(initialUri || '/');
    }

    replace(): void {}

    _update(url: string): void {
        const urlParts = new UrlParts(url);
        this.pathname = urlParts.getPath();
        this.search = urlParts.getQuery();
        this.hash = urlParts.getFragment();
        this.href = url;
    }
}
