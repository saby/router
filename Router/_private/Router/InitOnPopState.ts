import UrlRewriter from '../UrlRewriter';
import { IRouter } from './Router';
import { IHistoryState } from '../DataInterfaces';

/**
 * Подписка на событие "onpopstate" для отслеживания нажатия по стрелка перехода по истории браузера
 */
export default function initOnPopState(routerGetter: () => IRouter): void {
    if (typeof window === 'undefined') {
        return;
    }

    /**
     * при невозможности перейти к следующему состоянию, блокируем обработку события "popstate",
     * пока не выполнится errback метода IRouter.navigate
     */
    let skipNextChange: boolean = false;

    window.onpopstate = (event: PopStateEvent) => {
        if (skipNextChange) {
            skipNextChange = false;
            return;
        }

        // тут проверяем, что если в истории браузера записан какой-то объект состояния и в нем нет полей
        // роутера (id, state, href), то не вызываем navigate. Т.к. считаем, что кто-то записал состояние
        // в обход роутера и не хочет, чтобы совершали переход роутером
        if (event.state && !event.state.id && !event.state.state && !event.state.href) {
            return;
        }

        const router = routerGetter();

        const currentState: IHistoryState = router.history.getCurrentState();
        const prevState: IHistoryState = router.history.getPrevState();

        // нажатие на стрелку "назад" браузера
        if (
            (!event.state && !prevState) ||
            (event.state && event.state.id < (currentState.id as number))
        ) {
            const navigateToState: IHistoryState = _getNavigationState(
                prevState,
                event.state,
                event.state || prevState ? router.url.getStateUrl() : router.url.getUrl()
            );
            router.navigate(navigateToState, () => {
                return router.history.back(navigateToState);
            });
        } else {
            // нажатие на стрелку "вперед" браузера
            const nextState: IHistoryState = router.history.getNextState();
            const navigateToState: IHistoryState = _getNavigationState(
                nextState,
                event.state,
                router.url.getStateUrl()
            );
            router.navigate(
                navigateToState,
                () => {
                    return router.history.forward(navigateToState);
                },
                () => {
                    // unable to navigate to specified state, going back in history
                    skipNextChange = true;
                    window.history.back();
                }
            );
        }
    };
}

function _getNavigationState(
    localState: IHistoryState,
    windowState: IHistoryState,
    currentUrl: string
): IHistoryState {
    if (!localState) {
        if (windowState && windowState.state && windowState.href) {
            return windowState;
        }
        return {
            state: UrlRewriter.getInstance().get(currentUrl),
            href: currentUrl,
        };
    }
    if (windowState && windowState.state && windowState.href) {
        return windowState;
    }
    return localState;
}
