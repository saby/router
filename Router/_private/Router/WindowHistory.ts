import { IWindowLocation } from './WindowLocation';
import { IHistoryState } from '../DataInterfaces';

export interface IWindowHistory {
    /**
     * Признак того, что объект истории создан как заглушка. При открытии приложения в приложении
     */
    isFake?: boolean;
    state: IHistoryState;
    pushState(state: IHistoryState, title: string, url: string): void;
    replaceState(state: IHistoryState, title: string, url: string): void;
    back(): void;
}

/**
 * Класс-заглушка для window.history внутри роутера при работе роутера в приложении на панели
 * @private
 */
export default class WindowHistory implements IWindowHistory {
    isFake: boolean = true;
    private _states: IHistoryState[] = [];
    get state(): IHistoryState {
        return this._states[this._states.length - 1];
    }

    constructor(private _location: IWindowLocation) {}

    pushState(state: IHistoryState, _: string, url: string): void {
        this._states.push(state);
        // необходимо актуализировать location новым url. В window.history этим занимался бы браузер
        this._location._update(url);
    }

    replaceState(state: IHistoryState, _: string, url: string): void {
        this._states.pop();
        this._states.push(state);
        // необходимо актуализировать location новым url. В window.history этим занимался бы браузер
        this._location._update(url);
    }

    back(): void {
        this._states.pop();
        const state = this._states[this._states.length - 1];
        if (state) {
            // необходимо актуализировать location новым url. В window.history этим занимался бы браузер
            this._location._update(state.href || state.state);
        }
    }
}
