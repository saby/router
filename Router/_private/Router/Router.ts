import { getStore, location as envLocation, logger, ILocation } from 'Application/Env';
import UrlRewriter from '../UrlRewriter';
import RouterUrl from './RouterUrl';
import MaskResolver, { getAppNameByUrl } from '../MaskResolver';
import History from '../History';
import RouterManager from './RouterManager';
import WindowLocation, { IWindowLocation } from './WindowLocation';
import WindowHistory, { IWindowHistory } from './WindowHistory';
import { IHistoryState, TOnChangeHistoryState } from '../DataInterfaces';

/**
 * Интерфейс класса реализующего API работы с роутингом.
 * Более подробно API роутинга описано по {@link /doc/platform/developmentapl/interface-development/routing/#router-api ссылке}
 * @public
 * @author Мустафин Л.И.
 */
export interface IRouter {
    /**
     * Набор методов для работы с router.json, в котором можно задать соответствие
     * между текущим путем и его короткой записью - "красивым" URL
     */
    readonly urlRewriter: UrlRewriter;
    /**
     * Класс для работы с url в рамках Роутинга.
     */
    readonly url: RouterUrl;
    /**
     * Набор методов обеспечивающих работу с масками и параметрами URL
     */
    readonly maskResolver: MaskResolver;
    /**
     * Набор методов для взаимодействия с Историей Роутера
     */
    readonly history: History;
    /**
     * Менеджер обработчиков (до и после SPA перехода) для контролов Route и Reference.
     * @hidden
     */
    readonly _manager: RouterManager;
    /**
     * Производит single-page переход (без перезагрузки страницы) к новому состоянию (URL-адресу).
     * @param newState Состояние для перехода.
     * @param callback Необязательная функция, будет вызвана вместо window.history.pushState.
     * @param errback Необязательная функция, будет вызвана если один из компонентов Route отменит переход.
     *
     * @example
     * <pre>
     * import { IHistoryState } from 'Router/router';
     * // ...
     * _afterMount(): void {
     *     const state: IHistoryState = {state: 'some/new/state', href: 'pretty_url'}
     *     this._options.Router.navigate(state, () => true, (err) => alert(err.message));
     * }
     * // ...
     * </pre>
     *
     */
    navigate: (
        newState: IHistoryState,
        callback?: () => void | boolean,
        errback?: (err?: Error) => void
    ) => void;
    /**
     * Производит переход без перезагрузки страницы без добавления новой записи в историю переходов (вместо этого заменяет текущую).
     * @param newHistoryState Состояние для перехода.
     * @param callback Функция, которая будет вызвана после изменения URL адреса в адресной строке
     * @see Router/router:IRouter#navigate
     *
     * @example
     * <pre>
     * import { IHistoryState } from 'Router/router';
     * // ...
     * _afterMount(): void {
     *     const state: IHistoryState = {state: 'some/new/state'};
     *     this._options.Router.replaceState(state, (newState) => alert('Переход совершён!'));
     * }
     * // ...
     * </pre>
     */
    replaceState: (
        newHistoryState: IHistoryState,
        callback?: (rewrittenHistoryState: IHistoryState) => void
    ) => void;
}

/**
 * Класс реализующий API работы с роутингом
 * @private
 */
export default class Router implements IRouter {
    /**
     * Признак того, что уже идет процесс SPA перехода
     */
    private _isNavigating: boolean = false;
    protected _instId: number;

    constructor(
        readonly urlRewriter: UrlRewriter,
        readonly url: RouterUrl,
        readonly maskResolver: MaskResolver,
        readonly history: History,
        readonly _manager: RouterManager
    ) {
        this._instId = Router._instIdCounter++;
    }

    /**
     * Производит single-page переход (без перезагрузки страницы) к новому состоянию (URL-адресу).
     * @param newState Состояние для перехода.
     * @param callback Необязательная функция, будет вызвана вместо window.history.pushState.
     * @param errback Необязательная функция, будет вызвана если один из компонентов Route отменит переход.
     *
     * @example
     * <pre>
     * import { IHistoryState } from 'Router/router';
     * // ...
     * _afterMount(): void {
     *     const state: IHistoryState = {state: 'some/new/state', href: 'pretty_url'}
     *     this._options.Router.navigate(state, () => true, (err) => alert(err.message));
     * }
     * // ...
     * </pre>
     *
     */
    navigate(
        newState: IHistoryState,
        callback?: () => void | boolean,
        errback?: (err?: Error) => void
    ): void {
        const rewrittenNewUrl: string = this.urlRewriter.get(newState.state);
        const prettyUrl: string = newState.href || this.urlRewriter.getReverse(rewrittenNewUrl);
        const currentState: IHistoryState = this.history.getCurrentState();

        if (currentState.state === rewrittenNewUrl || this._isNavigating) {
            return;
        }
        const rewrittenNewState: IHistoryState = {
            state: rewrittenNewUrl,
            href: prettyUrl,
        };
        this._isNavigating = true;

        const tryApplyNewStateCallback = (accept: boolean) => {
            this._isNavigating = false;
            if (accept) {
                if (callback) {
                    if (callback() === true) {
                        this.history.push(rewrittenNewState);
                    }
                } else {
                    this.history.push(rewrittenNewState);
                }
                this._manager.callAfterUrlChange(rewrittenNewState, currentState);
            } else if (errback) {
                errback();
            }
        };
        const tryApplyNewStateErrback = (err: Error) => {
            this._isNavigating = false;
            if (errback) {
                errback(err);
            } else {
                logger.error('Ошибка при вызове IRouter.navigate - ' + err.message, err);
            }
        };

        let result: Promise<boolean> | boolean;
        try {
            result = this._tryApplyNewState(rewrittenNewState);
        } catch (err) {
            tryApplyNewStateErrback(err as Error);
            return;
        }

        if (result instanceof Promise) {
            result.then(tryApplyNewStateCallback, tryApplyNewStateErrback);
            return;
        }
        tryApplyNewStateCallback(result);
    }

    /**
     * Производит переход без перезагрузки страницы без добавления новой записи в историю переходов (вместо этого заменяет текущую).
     * @param newHistoryState Состояние для перехода.
     * @param callback Функция, которая будет вызвана после изменения URL адреса в адресной строке
     * @see Router/router:IRouter#navigate
     *
     * @example
     * <pre>
     * import { IHistoryState } from 'Router/router';
     * // ...
     * _afterMount(): void {
     *     const state: IHistoryState = {state: 'some/new/state'};
     *     this._options.Router.replaceState(state, (newState) => alert('Переход совершён!'));
     * }
     * // ...
     * </pre>
     */
    replaceState(
        newHistoryState: IHistoryState,
        callback?: (rewrittenHistoryState: IHistoryState) => void
    ): void {
        const rewrittenState: string = this.urlRewriter.get(newHistoryState.state);
        const rewrittenHref: string =
            newHistoryState.href || this.urlRewriter.getReverse(rewrittenState);
        const rewrittenHistoryState: IHistoryState = {
            state: rewrittenState,
            href: rewrittenHref,
        };

        this.navigate(rewrittenHistoryState, () => {
            this.history.replaceState(rewrittenHistoryState);
            callback?.(rewrittenHistoryState);
        });
    }

    private _tryApplyNewState(newState: IHistoryState): Promise<boolean> | boolean {
        const currentState: IHistoryState = this.history.getCurrentState();
        const newApp: string = getAppNameByUrl(newState.state);
        const currentApp: string = getAppNameByUrl(currentState.state);

        const callback = (res: boolean): boolean => {
            if (newApp === currentApp) {
                return res;
            }
            // Переходим без СПА, потому что сломалась синхронизация HEAD
            // при несовпадении VirtualDom inferno разрушает HEAD. Нужно
            // допатчить инферно так, чтобы под капотом библиотека НИКОГДА
            // не удаляла HEAD и HTML
            this.url.location.href = newState.href as string;
            return false;
        };

        const result: Promise<boolean> | boolean = this._manager.callBeforeUrlChange(
            newState,
            currentState
        );
        if (result instanceof Promise) {
            return result.then((res) => {
                return callback(res);
            });
        }
        return callback(result);
    }

    private static _instIdCounter: number = 0;
}

/**
 * Получить "корневой" объект Router,
 * который иницализирует все методы API роутера относительно window.location и window.history
 * @public
 */
export function getRootRouter(
    logCall: boolean = false,
    onChangeHistoryState?: TOnChangeHistoryState
): IRouter {
    if (logCall) {
        const error = new Error(
            'Вызов getRootRouter там, где это не нужно. \
            Необходимо перейти на использование API роутера из контекста.'
        );
        logger.warn(error);
    }
    const store = getStore<Record<string, IRouter>>('ROOT_ROUTER');
    let rootRouter = store.get('instance');
    if (!rootRouter) {
        rootRouter = _createRouter(
            envLocation,
            typeof window !== 'undefined'
                ? window.history
                : new WindowHistory(new WindowLocation()),
            onChangeHistoryState
        );
        store.set('instance', rootRouter);
    }
    return rootRouter;
}

/**
 * Создание нового объекта Router, который иницализирует все методы API роутера относительно нового "контекста"
 * Этот новый Router всегда будет работать с фейковым window.location и window.history
 * @param initialUri
 * @private
 */
export function createNewRouter(
    initialUri: string = '/',
    onChangeHistoryState?: TOnChangeHistoryState
): IRouter {
    const fakeLocation: IWindowLocation = new WindowLocation(initialUri);
    const fakeHistory: IWindowHistory = new WindowHistory(fakeLocation);
    return _createRouter(fakeLocation, fakeHistory, onChangeHistoryState);
}

function _createRouter(
    _location: ILocation,
    _history: IWindowHistory,
    onChangeHistoryState?: TOnChangeHistoryState
): IRouter {
    const urlRewriter = UrlRewriter.getInstance();
    const routerUrl = new RouterUrl(_location, urlRewriter);
    const maskResolver = new MaskResolver(urlRewriter, routerUrl);
    const routerHistory = new History(urlRewriter, routerUrl, _history, onChangeHistoryState);
    const routerManager = new RouterManager();
    return new Router(urlRewriter, routerUrl, maskResolver, routerHistory, routerManager);
}
