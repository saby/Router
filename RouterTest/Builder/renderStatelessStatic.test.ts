import { renderStatic } from 'Router/Builder';

describe('renderStatelessStatic', () => {
    test('main', () => {
        // @ts-ignore
        const html = renderStatic({
            wsRoot: '%{WI.SBIS_ROOT}',
            resourceRoot: '%{RESOURCE_ROOT}',
            metaRoot: '%{META_ROOT}',
            defaultServiceUrl: '%{SERVICES_PATH}',
            appRoot: '%{APPLICATION_ROOT}',
            // @ts-ignore
            RUMEnabled: '%{RUM_ENABLED}',
            pageName: '%{PAGE_NAME}',
            builder: 'function anonymous() {return"";}',
            buildStateless: true,
            dependencies: ['Module/Index'],
        });
        expect(html).toMatchSnapshot();
    });
});
