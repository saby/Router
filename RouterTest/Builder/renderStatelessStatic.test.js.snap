// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderStatelessStatic main 1`] = `
"<!DOCTYPE html>
<html lang=ru>
  <head>
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\" sid=\\"h-ps-1\\">
<meta charset=\\"utf-8\\" class=\\"head-server-block\\" sid=\\"h-ps-2\\">
<script sid=\\"h-ps-3\\" type=\\"text/javascript\\">function getQP() {
    var sD = [\\"wsRoot\\", \\"resourceRoot\\", \\"metaRoot\\", \\"defaultServiceUrl\\", \\"appRoot\\", \\"staticDomains\\", \\"product\\", \\"cdnRoot\\"];
    var r = {};
    var f = [];
    var items = location.search.substring(1).split(\\"&\\");
    for (var index = 0; index < items.length; index++) {
        f = items[index].split(\\"=\\");
        if (sD.indexOf(f[0]) !== -1) {
            var decoded = decodeURIComponent(f[1]);
            r[f[0]] = decoded === \\"staticDomains\\" ? JSON.parse(decoded) : decoded;
        }
    }
    return r;
}
var cfg = window.wsConfig = window.wsConfig || getQP();
cfg.wsRoot = cfg.wsRoot || '/resources/WS.Core';
cfg.resourceRoot = cfg.resourceRoot || '/resources/';
cfg.metaRoot = cfg.metaRoot || '/resources/';
cfg.cdnRoot = cfg.cdnRoot || '/cdn/';
cfg.defaultServiceUrl = cfg.defaultServiceUrl || '/service/';
cfg.appRoot = cfg.appRoot || '/';
cfg.RUMEnabled = false;
cfg.pageName = '';
cfg.userConfigSupport = true;
cfg.trackErrors = true;
cfg.staticDomains = cfg.staticDomains || [];
cfg.compatible = false;
cfg.product = cfg.product || 'undefined';
cfg.xDomainSystemname = '';
window.buildnumber = cfg.buildnumber || undefined;
window['X-UNIQ-ID'] = '';
window['X-REQUESTUUID'] = '';
window['X-CURRENTMETHOD'] = '';
window['csrStartTime'] = Date.now();
function onErrorHandler(name) { window.wsConfig.loadingStatus = window.wsConfig.loadingStatus || {}; window.wsConfig.loadingStatus[name] = 'ERROR'; try { if (window.sessionStorage.getItem('theForceReloadWasEarly')) { window.wsConfig.loadingStatus.theForceReloadWasEarly = true; if (!window.wsConfig.loadingStatus.dialogShown) { setTimeout(function () { var res = window.confirm('При загрузке страницы не все ресурсы были загружены.' +' Страница может работать некорректно. Перезагрузить страницу?'); if (res) { window.location.reload(); } }, 100); window.wsConfig.loadingStatus.dialogShown = true; } return; } window.sessionStorage.setItem('theForceReloadWasEarly', 'true'); window.location.reload(); } catch(err) { /* sessionStorage недоступен */} };
window.themeName = '';</script>
  </head>
  <body >
    <div id=\\"wasaby-content\\" style=\\"width: inherit; height: inherit;\\" application=\\"Module/Index\\">
      <div></div>
    </div>
    <div class=\\"wasabyBaseDeps\\">
      <script type=\\"text/javascript\\" key=\\"require\\" src=\\"/cdn/RequireJS/2.3.5-p10/require.min.js\\" onerror=\\"onErrorHandler('require')\\" sid=\\"s-1\\" defer=\\"defer\\"></script>
    </div>


    <div id=\\"wasabyStartScript\\">
      <script key=\\"init_script\\">        document.addEventListener('DOMContentLoaded', function () {            var wasabyBaseDeps = document.getElementsByClassName('wasabyBaseDeps')[0];            var scripts = [['bundles',window.wsConfig.metaRoot+'bundles.min.js'],['contents',window.wsConfig.metaRoot+'contents.min.js'],['router',window.wsConfig.metaRoot+'router.min.js'],];            var promises = [];            for (var scr of scripts) {                var prms = new Promise((resolve, reject) => {                    var addScript = document.createElement('script');                    addScript.src = scr[1];                    addScript.key = scr[0];                    addScript.onload = function () {                        resolve();                    };                    addScript.onerror = function (event) {                        onErrorHandler(event.target.key);                        reject();                    };                    wasabyBaseDeps.appendChild(addScript);                });                promises.push(prms);            }            Promise.all(promises).then(function () {                /* buildnumber можем достать только из contents.js */                var bNumber = window.contents.buildnumber;                window.wsConfig.buildnumber = bNumber;                window.buildnumber = bNumber;                if (window.contents && window.contents.modules && window.contents.modules.RequireJsLoader && window.contents.modules.RequireJsLoader.buildnumber) {                    bNumber = window.contents.modules.RequireJsLoader.buildnumber;                }                var requireConfigPromise = new Promise((resolve, reject) => {                    var addScript = document.createElement('script');                    addScript.src = window.wsConfig.resourceRoot+'RequireJsLoader/config.min.js?x_module=' + bNumber;                    addScript.key = 'config';                    addScript.onload = function () {                        resolve();                    };                    addScript.onerror = function (event) {                        onErrorHandler(event.target.key);                        reject();                    };                    wasabyBaseDeps.appendChild(addScript);                });                requireConfigPromise.then(function () {                                require(['Env/Env', 'Application/Initializer', 'Application/Env', 'SbisUI/Wasaby', 'UI/Base', 'UI/State', 'Application/State', 'Router/router', 'SbisUI/polyfill'],            function(Env, AppInit, AppEnv, EnvUIWasaby, UIBase, UIState, AppState, router){                Object.assign(Env.constants, window.wsConfig);                require(['WasabyLoader/ModulesLoader','Module/Index'], function(ModulesLoader){                    if (performance && performance.mark) {                        performance.mark('SCRIPTS COMPILING END');                        performance.mark('CORE INIT START');                    }                    var sr = new AppState.StateReceiver(UIState.Serializer);                    AppInit.default(window.wsConfig, void 0, sr);                    var calcMethodName = function(url, maskResolver) {            if (url === '/') {                return 'PSWaSabyRouting.index_html';            }            var params = maskResolver.calculateUrlParams('/:component/:pageId', url);            if (params.component === 'page') {                /* pagex */                return 'PSWaSabyRouting.' + params.pageId;            }            var hyphen = params.component ? '-' : '';            return 'PSWaSabyRouting.' + params.component + hyphen + 'Index';        };        var onChangeState = function(newState) {            if (newState.spaHistory) { window.spaHistory = newState.spaHistory; }            if (newState.href && Router) {                var pageName = calcMethodName(newState.href, Router.maskResolver);                window['X-CURRENTMETHOD'] = pageName;                Env.constants.pageName = pageName;            };        };                    var Router = router.getRootRouter(false, onChangeState);                    UIBase.BootstrapStart({ Router: Router }, document.getElementById('wasaby-content'));                    ModulesLoader.initWarmup();                    try {                        window.sessionStorage.removeItem('theForceReloadWasEarly');                    } catch(err) { /* sessionStorage недоступен */}                    if (performance && performance.mark) {                        performance.mark('CORE INIT END');                    }                });                            });                });            });        });        </script>
    </div>
<div id=\\"sbisEnvUI_errorContainer\\"></div>
<script async src=\\"/cdn/Maintenance/1.0.41/js/checkSoftware.min.js\\"></script>
  </body>
</html>"
`;
